ledger-service/fetch-contracts/src/main/scala/fetchcontracts/domain/ContractTypeId.scala

```diff
@@ -59,6 +61,75 @@ sealed abstract class ContractTypeId[+PkgId]
   }
 }
 
+object ResolvedQuery {
+  def apply(resolved: ContractTypeId.Resolved): ResolvedQuery = {
+    import ContractTypeId._
+    resolved match {
+      case t @ Template(_, _, _) => ByTemplateId(t)
+      case i @ Interface(_, _, _) => ByInterfaceId(i)
+    }
+  }
+
+  def apply(resolved: Set[ContractTypeId.Resolved]): Unsupported \/ ResolvedQuery = {
+    val (templateIds, interfaceIds) = partition(resolved)
+    if (templateIds.nonEmpty && interfaceIds.nonEmpty) {
+      -\/(CannotQueryBothTemplateIdsAndInterfaceIds)
+    } else if (templateIds.isEmpty && interfaceIds.size != 1) {
+      -\/(CannotQueryManyInterfaceIds)
+    } else if (templateIds.isEmpty && interfaceIds.size == 1) {
+      \/-(ByInterfaceId(interfaceIds.head))
+    } else if (templateIds.nonEmpty) {
+      \/-(ByTemplateIds(templateIds))
+    } else {
+      -\/(CannotBeEmpty)
```

Comment:
It's not immediately apparent, but this is unreachable because the second branch matches this case.

While you don't have `unapplySeq`s provided for `Set`, it might be nice to define an `EmptySet { def unapply(s: Set[_]): Boolean = s.isEmpty }` pattern and use `NonEmpty` and perhaps `Single` from the `nonempty` package to declare the possible patterns.

Reply:
Good catch (and broke the CI), fixed now.

Reply:
We can revisit ResolvedQuery and improve it more over time, I would imagine this type will grow a little further, getting rid of the empty case, and maybe unify it with some tracking of user intent and response based on that (if that is needed, not sure)